source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "source/extensions/config_subscription/grpc/grpc_stream.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * This class arbitrates between two config providers of the same GrpcMux -
+ * the primary and the failover. Envoy always prefers fetching config from the
+ * primary source, but if not available, will fetch the config from the failover
+ * source until the primary is again available.
+ *
+ * This class owns the state for the GrpcMux primary and failover streams, and
+ * proxies the gRPC stream functionality to either the primary or the failover config sources.
+ * The failover source is optional and will only be used if passed upon initialization.
+ *
+ * The primary config source is always preferred over the failover source. If the primary
+ * is available, the stream to the failover source will terminated.
+ *
+ * Failover is supported in both
+ * SotW (RequestType = envoy::service::discovery::v3::DiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
+ *
+ * The use of this class will be as follows: the GrpcMux object will own an instance of
+ * the GrpcMuxFailoverManager. The GrpcMuxFailoverManager will own 2 GrpcStreams, primary
+ * and secondary. Each of the primary and secondary streams will invoke GrpcStreamCallbacks
+ * on their corresponding objects (also owned by the GrpcMuxFailoverProxy). These invocations
+ * will be followed by the GrpcMuxFailoverProxy calling the GrpcStreamCallbacks on the GrpcMux
+ * object that initialized it.
+ *
+ * Note: this class is WIP.
+ */
+template <class RequestType, class ResponseType> class GrpcMuxFailover {
```

Comment:
It seems all the interfaces are coming from GrpcInterface, so why we don't inherit from GrpcInterface? the GrpcMux will aware the failover logic?

Reply:
I was thinking about this as well, but decided against that.
My main objection was because typically inheritance implies an "is-a" relationship, and I don't know if `GrpcMuxFailover` is a GrpcStreamInterface...
That said, I think that this can always be changed later.
If you insist, I can add it now, but will leave it up to you.

Reply:
agree with that, we can change that later when we saw the full picture in the future